### ISSUE TITLE
Contextual typing fails when discriminant is an interpolated template literal

### ISSUE THREAD

**Issue by jcalz** [microsoft/TypeScript#53888]
# Bug Report

### 🔎 Search Terms

discriminated union, discriminant, interpolated template literal, contextual typing, delayed, deferred

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about discriminated unions

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ssl=9&ssc=1&pln=8&pc=47#code/C4TwDgpgBAylC8UDeUAmAuKAiAzlg3FAMYBGmAFAB6Y7ABOAlgHYDmAlAgHxQBuA9g1SEAvvgBQoSFAByCZGkxYmBYmShVMTAK4BbEhDod43foJHjUEIgBsAhnWgAzLUyLAGfJlEd8+5H3yYcAA+MmyYpkJiAeQoGNjKADSqmJRcUJQAdMB8AGIMlBCo5BzCbIQA9BVQfADWtiDRvrEK2HjJpKnpWTkAqmCQdADCtjgQJVBlldV1DVBNfnGYAAYAJEi4WMLLHWppxhnZfAAapxNTUFVQBnR8dACEYlcAtK9v7x9v3AB+T9UACvZbDoIMADFAAOSUCFQBg6MDWBhEBjAawgKAAC1GUFsXghuJAMMkEEyYiAA)

### 💻 Code

```ts
type S = { d: "s"; cb: (x: string) => void; };
type N = { d: "n"; cb: (x: number) => void; };
declare function foo(foo: S | N): void;
foo({ d: "n", cb: x => x.toFixed() }); // okay
foo({ d: "s", cb: x => x.toUpperCase() }); // okay 
foo({ d: `${"s"}`, cb: x => x.toXXX() }); // error!
// ------------------> ~
// Parameter 'x' implicitly has an 'any' type.
```

### 🙁 Actual behavior

In the third call to `foo()`, the discriminant property `d` is an interpolated template literal whose type is correctly seen to be `"s"`, but the `cb` property's callback parameter `x` is not contextually typed as `string` and instead implicitly falls back to `any`.  

### 🙂 Expected behavior

The third call to `foo()` should behave just like the second call to `foo()`, where the object is narrowed to `S` and therefore `cb`'s parameter is contextually typed as `string`.

### Notes

Comes from [this Stack Overflow question](https://stackoverflow.com/q/76031041/2887218)

I'm imagining this is just a design limitation where the type of `d` is computed too late for it to be available when the contextual typing for `cb` happens.  A similar failure occurs when the discriminant is the output of an inline function call:

```ts
function justS(): "s" { return "s" };
foo({ d: justS(), cb: x => x.toXXX() }); // error!
// -----------------> ~
// Parameter 'x' implicitly has an 'any' type.
```
[Playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAylC8UDeUAmAuKAiAzlg3FAMYBGmAFAB6Y7ABOAlgHYDmAlAgHxQBuA9g1SEAvvgBQoSFAByCZGkxYmBYmShVMTAK4BbEhDod43foJHjUEIgBsAhnWgAzLUyLAGfJlEd8+5H3yYcAA+MmyYpkJiAeQoGNjKADSqmJRcUJQAdMB8AGIMlBCo5BzCbIQA9BVQfADWtiDRvrEK2HjJpKnpWTkAqmCQdADCtjgQJVBlldV1DVDRLm4eXgBWWrQwJYp48g7AWnReuFiT4jFxmGsbJR1qacYZ2XwAGq8TU1BVUAZ0fHQAhGIvgBaUFg8EQ0HcAB+QOqAAV7LYdBBgAYoABySgYqAMHRgawMIgMYDWEBQAAWoygti8GNpIBxkggmSAA)

I searched around for an existing issue but I didn't find an exact enough match.  There's #35769 / #41759 / #46847, for example... related but not obviously the same.

The obvious workaround is just to do the interpolation ahead of time into a `const` and use that instead:

```ts
const s = `${"s"}` as const;
foo({ d: s, cb: x => x.toUpperCase() }); // okay
```
[Playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAylC8UDeUAmAuKAiAzlg3FAMYBGmAFAB6Y7ABOAlgHYDmAlAgHxQBuA9g1SEAvvgBQoSFAByCZGkxYmBYmShVMTAK4BbEhDod43foJHjUEIgBsAhnWgAzLUyLAGfJlEd8+5H3yYcAA+MmyYpkJiAeQoGNjKADSqmJRcUJQAdMB8AGIMlBCo5BzCbIQA9BVQfADWtiDRvrEK2HjJpKnpWTkAqmCQdADCtjgQJVBlldV1DVBiRJ60UDhyAAYAJEi4WMJrUKPES8DiMXE0HWppxhnZfP2DI2MTU1BVNfWNQA)

Mostly I'm just looking for an official status on this (and expecting it to be Design Limitation).

**Comment by fatcerberus** [comment]
This feels like one of those “you were trying to break it and succeeded” type deals that will likely not be considered worth fixing since it’s unlikely anyone would hit it by accident and/or be blocked by it (i.e. “Why would you write it this way?”)

**Comment by fatcerberus** [comment]
Hmm, while it doesn't invalidate the OP here, that SO question has this type:
```ts
type BiomePlainLinkProps = {
  href: string;
  onClick?: (event: string) => void;
}

type BiomeButtonProps = {
  href?: never;
  onClick?: (event: number) => void;
}

type ClickableDiscriminatedUnion =
  | BiomePlainLinkProps
  | BiomeButtonProps;
```

I'm not even sure that's a valid discriminated union?  I thought the discriminant had to be a literal type (or at least a union of them), not just `string`.

**Comment by jcalz** [comment]
The discriminant there is an optional `undefined`
